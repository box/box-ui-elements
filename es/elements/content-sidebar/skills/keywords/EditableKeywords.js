function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
/**
 * 
 * @file Editable Skill Keywords card component
 * @author Box
 */

import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import PillSelector from '../../../../components/pill-selector-dropdown/PillSelector';
import PrimaryButton from '../../../../components/primary-button/PrimaryButton';
import Button from '../../../../components/button/Button';
import messages from '../../../common/messages';
import { SKILLS_TARGETS } from '../../../common/interactionTargets';
import getPills from './keywordUtils';
import './EditableKeywords.scss';
class EditableKeywords extends React.PureComponent {
  /**
   * [constructor]
   *
   * @public
   * @return {EditableKeywords}
   */
  constructor(props) {
    super(props);
    /**
     * Called when keywords gets new properties.
     * Should reset to original state.
     *
     * @private
     * @param {Object} option - pill
     * @param {number} index - pill index
     * @return {void}
     */
    _defineProperty(this, "onRemove", (option, index) => {
      // eslint-disable-line
      const {
        onDelete,
        keywords
      } = this.props;
      onDelete(keywords[index]);
    });
    /**
     * When pressing enter in the pill input box
     *
     * @private
     * @param {Event} event - keyboard event
     * @return {void}
     */
    _defineProperty(this, "onKeyDown", ({
      key
    }) => {
      if (key === 'Enter' && !this.state.isInCompositionMode) {
        this.onBlur();
      }
    });
    /**
     * Called when pill selector is blurred.
     * Adds a new pill if needed.
     *
     * @private
     * @return {void}
     */
    _defineProperty(this, "onBlur", () => {
      const {
        onAdd
      } = this.props;
      const {
        keyword
      } = this.state;
      if (keyword) {
        onAdd({
          type: 'text',
          text: keyword
        });
      }
    });
    /**
     * Enables composition mode.
     *
     * @private
     * @return {void}
     */
    _defineProperty(this, "onCompositionStart", () => {
      this.setState({
        isInCompositionMode: true
      });
    });
    /**
     * Disables composition mode.
     *
     * @private
     * @return {void}
     */
    _defineProperty(this, "onCompositionEnd", () => {
      this.setState({
        isInCompositionMode: false
      });
    });
    /**
     * Called when pill selector gets new input value.
     *
     * @private
     * @return {void}
     */
    _defineProperty(this, "onInput", event => {
      const currentTarget = event.currentTarget;
      this.setState({
        keyword: currentTarget.value
      });
    });
    this.state = {
      pills: getPills(props.keywords),
      keyword: '',
      isInCompositionMode: false
    };
  }

  /**
   * Called when keywords gets new properties.
   * Should reset to original state.
   *
   * @private
   * @param {Object} nextProps - component props
   * @return {void}
   */
  componentDidUpdate({
    keywords: prevKeywords
  }) {
    const {
      keywords
    } = this.props;
    if (prevKeywords !== keywords) {
      this.setState({
        pills: getPills(keywords),
        keyword: ''
      });
    }
  }
  /**
   * Renders the keywords
   *
   * @private
   * @return {void}
   */
  render() {
    const {
      onSave,
      onCancel
    } = this.props;
    const {
      pills,
      keyword
    } = this.state;
    return /*#__PURE__*/React.createElement("span", {
      className: "bdl-EditableKeywords"
    }, /*#__PURE__*/React.createElement(PillSelector, {
      onBlur: this.onBlur,
      onCompositionEnd: this.onCompositionEnd,
      onCompositionStart: this.onCompositionStart,
      onInput: this.onInput,
      onKeyDown: this.onKeyDown,
      onPaste: this.onInput,
      onRemove: this.onRemove,
      selectedOptions: pills,
      value: keyword
    }), /*#__PURE__*/React.createElement("div", {
      className: "be-keywords-buttons"
    }, /*#__PURE__*/React.createElement(Button, {
      "data-resin-target": SKILLS_TARGETS.KEYWORDS.EDIT_CANCEL,
      onClick: onCancel,
      type: "button"
    }, /*#__PURE__*/React.createElement(FormattedMessage, messages.cancel)), /*#__PURE__*/React.createElement(PrimaryButton, {
      "data-resin-target": SKILLS_TARGETS.KEYWORDS.EDIT_SAVE,
      onClick: onSave,
      type: "button"
    }, /*#__PURE__*/React.createElement(FormattedMessage, messages.save))));
  }
}
export default EditableKeywords;
//# sourceMappingURL=EditableKeywords.js.map