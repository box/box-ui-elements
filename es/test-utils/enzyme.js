function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
/* eslint-disable import/no-extraneous-dependencies */
import * as React from 'react';
import { mount as baseMount, shallow as baseShallow } from 'enzyme';

// Global providers that need setup
import { ThemeProvider } from 'styled-components';
import defaultTheme from '../styles/theme';

// Wrap this around every component so they can use app-wide context.
// Prevents breaking due to missing providers.
// See https://github.com/airbnb/enzyme/blob/master/docs/api/ReactWrapper/getWrappingComponent.md
/* eslint-disable react/prop-types */
const Wrappers = ({
  children
}) => {
  return /*#__PURE__*/React.createElement(ThemeProvider, {
    theme: defaultTheme
  }, children);
};

/**
 * mount() from Enzyme but with required app-wide context providers included by default
 */
const mountConnected = (element, options = {}) => baseMount(element, _objectSpread({
  wrappingComponent: Wrappers
}, options));

/**
 * shallow() from Enzyme but with required app-wide context providers included by default.
 */
const shallowConnected = (element, options = {}) => baseShallow(element, _objectSpread({
  wrappingComponent: Wrappers
}, options));

// Re-export everything from Enzyme
export * from 'enzyme';
// Export wrapped methods
export { mountConnected, shallowConnected };
//# sourceMappingURL=enzyme.js.map