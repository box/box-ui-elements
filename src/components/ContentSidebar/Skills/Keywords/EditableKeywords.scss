@import '../../../variables';
@import '~box-react-ui/lib/styles/common/pills';
@import '~box-react-ui/lib/components/pill-selector-dropdown/PillSelector';

.be-keywords {
    .be-keywords-buttons {
        margin: 7px -5px 0 0;
        text-align: right;

        .btn {
            margin: 0 0 0 5px;
        }
    }

    .pill-selector-wrapper {
        margin: 0;

        .pill-selector-input-wrapper {
            margin: 0;
            width: 100%;

            input[type=text].pill-selector-input {
                border: 1px solid $almost-white;
                box-shadow: none;
                height: 24px;
                margin: 2px 0 0 2px;
                min-width: 97%;
                padding: 0;
            }
        }
    }
}
