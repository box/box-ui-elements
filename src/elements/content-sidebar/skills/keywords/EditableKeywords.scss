@import '../../../common/variables';

.be-keywords {
    .be-keywords-buttons {
        margin: 7px -5px 0 0;
        text-align: right;

        .btn {
            margin: 0 0 0 5px;
        }
    }

    .bdl-EditableKeywords {
        margin: 0;

        .bdl-PillSelector {
            width: 100%;
            margin: 0;

            input[type='text'].bdl-PillSelector-input {
                min-width: 97%;
                height: 24px;
                margin: 2px 0 0 2px;
                padding: 0;
                border: 1px solid $almost-white;
                box-shadow: none;
            }
        }
    }
}
