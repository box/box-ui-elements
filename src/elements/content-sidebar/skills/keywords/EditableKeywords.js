/**
 * @flow
 * @file Editable Skill Keywords card component
 * @author Box
 */

import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import PillSelector from '../../../../components/pill-selector-dropdown/PillSelector';
import PrimaryButton from '../../../../components/primary-button/PrimaryButton';
import Button from '../../../../components/button/Button';
import messages from '../../../common/messages';
import { SKILLS_TARGETS } from '../../../common/interactionTargets';
import getPills from './keywordUtils';
import type { Pill, Pills } from './flowTypes';
import type { SkillCardEntry } from '../../../../common/types/skills';
import './EditableKeywords.scss';

type Props = {
    keywords: Array<SkillCardEntry>,
    onAdd: Function,
    onCancel: Function,
    onDelete: Function,
    onSave: Function,
};

type State = {
    isInCompositionMode: boolean,
    keyword: string,
    pills: Pills,
};

class EditableKeywords extends React.PureComponent<Props, State> {
    props: Props;

    state: State;

    /**
     * [constructor]
     *
     * @public
     * @return {EditableKeywords}
     */
    constructor(props: Props) {
        super(props);
        this.state = {
            pills: getPills(props.keywords),
            keyword: '',
            isInCompositionMode: false,
        };
    }

    /**
     * Called when keywords gets new properties.
     * Should reset to original state.
     *
     * @private
     * @param {Object} nextProps - component props
     * @return {void}
     */
    componentDidUpdate({ keywords: prevKeywords }: Props): void {
        const { keywords } = this.props;

        if (prevKeywords !== keywords) {
            this.setState({ pills: getPills(keywords), keyword: '' });
        }
    }

    /**
     * Called when keywords gets new properties.
     * Should reset to original state.
     *
     * @private
     * @param {Object} option - pill
     * @param {number} index - pill index
     * @return {void}
     */
    onRemove = (option: Pill, index: number): void => {
        // eslint-disable-line
        const { onDelete, keywords }: Props = this.props;
        onDelete(keywords[index]);
    };

    /**
     * When pressing enter in the pill input box
     *
     * @private
     * @param {Event} event - keyboard event
     * @return {void}
     */
    onKeyDown = ({ key }: SyntheticKeyboardEvent<HTMLInputElement>): void => {
        if (key === 'Enter' && !this.state.isInCompositionMode) {
            this.onBlur();
        }
    };

    /**
     * Called when pill selector is blurred.
     * Adds a new pill if needed.
     *
     * @private
     * @return {void}
     */
    onBlur = () => {
        const { onAdd }: Props = this.props;
        const { keyword } = this.state;

        if (keyword) {
            onAdd({
                type: 'text',
                text: keyword,
            });
        }
    };

    /**
     * Enables composition mode.
     *
     * @private
     * @return {void}
     */
    onCompositionStart = () => {
        this.setState({ isInCompositionMode: true });
    };

    /**
     * Disables composition mode.
     *
     * @private
     * @return {void}
     */
    onCompositionEnd = () => {
        this.setState({ isInCompositionMode: false });
    };

    /**
     * Called when pill selector gets new input value.
     *
     * @private
     * @return {void}
     */
    onInput = (event: SyntheticKeyboardEvent<HTMLInputElement>) => {
        const currentTarget = (event.currentTarget: HTMLInputElement);
        this.setState({
            keyword: currentTarget.value,
        });
    };

    /**
     * Renders the keywords
     *
     * @private
     * @return {void}
     */
    render() {
        const { onSave, onCancel }: Props = this.props;
        const { pills, keyword }: State = this.state;
        return (
            <span className="bdl-EditableKeywords">
                <PillSelector
                    onBlur={this.onBlur}
                    onCompositionEnd={this.onCompositionEnd}
                    onCompositionStart={this.onCompositionStart}
                    onInput={this.onInput}
                    onKeyDown={this.onKeyDown}
                    onPaste={this.onInput}
                    onRemove={this.onRemove}
                    selectedOptions={pills}
                    value={keyword}
                />
                <div className="be-keywords-buttons">
                    <Button data-resin-target={SKILLS_TARGETS.KEYWORDS.EDIT_CANCEL} onClick={onCancel} type="button">
                        <FormattedMessage {...messages.cancel} />
                    </Button>
                    <PrimaryButton data-resin-target={SKILLS_TARGETS.KEYWORDS.EDIT_SAVE} onClick={onSave} type="button">
                        <FormattedMessage {...messages.save} />
                    </PrimaryButton>
                </div>
            </span>
        );
    }
}

export default EditableKeywords;
