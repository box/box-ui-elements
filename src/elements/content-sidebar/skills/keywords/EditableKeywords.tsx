import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import PillSelector from '../../../../components/pill-selector-dropdown/PillSelector';
import PrimaryButton from '../../../../components/primary-button/PrimaryButton';
import Button, { ButtonType } from '../../../../components/button/Button';
import messages from '../../../common/messages';
import { SKILLS_TARGETS } from '../../../common/interactionTargets';
import getPills from './keywordUtils';
import { Pill, Pills } from './types';
import { SkillCardEntry } from '../../../../common/types/skills';
import './EditableKeywords.scss';

interface Props {
    keywords: Array<SkillCardEntry>;
    onAdd: (keyword: SkillCardEntry) => void;
    onCancel: () => void;
    onDelete: (keyword: SkillCardEntry) => void;
    onSave: () => void;
}

interface State {
    isInCompositionMode: boolean;
    keyword: string;
    pills: Pills;
}

class EditableKeywords extends React.PureComponent<Props, State> {
    constructor(props: Props) {
        super(props);
        this.state = {
            pills: getPills(props.keywords),
            keyword: '',
            isInCompositionMode: false,
        };
    }

    componentDidUpdate({ keywords: prevKeywords }: Props): void {
        const { keywords } = this.props;

        if (prevKeywords !== keywords) {
            this.setState({ pills: getPills(keywords), keyword: '' });
        }
    }

    onRemove = (option: Pill, index: number): void => {
        const { onDelete, keywords } = this.props;
        onDelete(keywords[index]);
    };

    onKeyDown = (event: React.KeyboardEvent<HTMLInputElement>): void => {
        if (event.key === 'Enter' && !this.state.isInCompositionMode) {
            event.preventDefault();
            this.onBlur();
        }
    };

    onBlur = (): void => {
        const { onAdd } = this.props;
        const { keyword } = this.state;

        if (keyword) {
            onAdd({
                type: 'text',
                text: keyword,
            });
            this.setState({ keyword: '' });
        }
    };

    onCompositionStart = (): void => {
        this.setState({ isInCompositionMode: true });
    };

    onCompositionEnd = (): void => {
        this.setState({ isInCompositionMode: false });
    };

    onInput = (event: React.KeyboardEvent<HTMLInputElement>): void => {
        this.setState({
            keyword: event.currentTarget.value,
        });
    };

    render(): React.ReactElement {
        const { onSave, onCancel } = this.props;
        const { pills, keyword } = this.state;
        return (
            <span className="bdl-EditableKeywords">
                <PillSelector
                    onBlur={this.onBlur}
                    onCompositionEnd={this.onCompositionEnd}
                    onCompositionStart={this.onCompositionStart}
                    onInput={this.onInput}
                    onKeyDown={this.onKeyDown}
                    onPaste={this.onInput}
                    onRemove={this.onRemove}
                    selectedOptions={pills}
                    value={keyword}
                />
                <div className="be-keywords-buttons">
                    <Button
                        data-resin-target={SKILLS_TARGETS.KEYWORDS.EDIT_CANCEL}
                        onClick={onCancel}
                        type={ButtonType.BUTTON}
                    >
                        <FormattedMessage {...messages.cancel} />
                    </Button>
                    <PrimaryButton
                        data-resin-target={SKILLS_TARGETS.KEYWORDS.EDIT_SAVE}
                        onClick={onSave}
                        type={ButtonType.BUTTON}
                    >
                        <FormattedMessage {...messages.save} />
                    </PrimaryButton>
                </div>
            </span>
        );
    }
}

export default EditableKeywords;
