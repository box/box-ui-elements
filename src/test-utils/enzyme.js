/* eslint-disable import/no-extraneous-dependencies */
import * as React from 'react';
import { mount as baseMount, shallow as baseShallow } from 'enzyme';

// Global providers that need setup
import { ThemeProvider } from 'styled-components';
import defaultTheme from '../styles/theme';

// Wrap this around every component so they can use app-wide context.
// Prevents breaking due to missing providers.
// See https://github.com/airbnb/enzyme/blob/master/docs/api/ReactWrapper/getWrappingComponent.md
/* eslint-disable react/prop-types */
const Wrappers = ({ children }) => {
    return <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>;
};

/**
 * mount() from Enzyme but with required app-wide context providers included by default
 */
const mountConnected = (element, options = {}) => baseMount(element, { wrappingComponent: Wrappers, ...options });

/**
 * shallow() from Enzyme but with required app-wide context providers included by default.
 */
const shallowConnected = (element, options = {}) => baseShallow(element, { wrappingComponent: Wrappers, ...options });

// Re-export everything from Enzyme
export * from 'enzyme';
// Export wrapped methods
export { mountConnected, shallowConnected };
