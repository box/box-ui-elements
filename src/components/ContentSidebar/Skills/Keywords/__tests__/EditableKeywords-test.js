import * as React from 'react';
import { shallow } from 'enzyme';
import EditableKeywords from '../EditableKeywords';

describe('components/ContentSidebar/Skills/Keywords/EditableKeywords', () => {
    test('should correctly render', () => {
        const props = {
            keywords: [{ text: 'foo' }, { text: 'bar' }],
            onAdd: jest.fn(),
            onDelete: jest.fn(),
            onSave: jest.fn(),
            onCancel: jest.fn()
        };

        const wrapper = shallow(<EditableKeywords {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    describe('onKeyDown()', () => {
        test('should call onBlur when enter is pressed and is not in composition mode', () => {
            const wrapper = shallow(<EditableKeywords />);
            const instance = wrapper.instance();
            instance.onBlur = jest.fn();
            instance.onKeyDown({ key: 'Enter' });
            expect(instance.onBlur).toBeCalled();
        });
        test('should not call onBlur when in composition mode', () => {
            const wrapper = shallow(<EditableKeywords />);
            const instance = wrapper.instance();
            instance.setState({ isInCompositionMode: true });
            instance.onBlur = jest.fn();
            instance.onKeyDown({ key: 'Enter' });
            expect(instance.onBlur).not.toBeCalled();
        });
    });
});
