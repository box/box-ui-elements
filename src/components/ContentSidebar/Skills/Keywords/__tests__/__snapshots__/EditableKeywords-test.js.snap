// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ContentSidebar/Skills/Keywords/EditableKeywords should correctly render 1`] = `
<span
  className="pill-selector-wrapper"
>
  <PillSelector
    error=""
    inputProps={Object {}}
    onBlur={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    onRemove={[Function]}
    placeholder=""
    selectedOptions={
      Array [
        Object {
          "text": "foo",
          "value": 0,
        },
        Object {
          "text": "bar",
          "value": 1,
        },
      ]
    }
    value=""
  />
  <div
    className="be-keywords-buttons"
  >
    <Button
      onClick={[MockFunction]}
      type="button"
    >
      <FormattedMessage
        data-resin-target="skill-keywordeditcancel"
        defaultMessage="Cancel"
        id="be.cancel"
      />
    </Button>
    <PrimaryButton
      onClick={[MockFunction]}
      type="button"
    >
      <FormattedMessage
        data-resin-target="skill-keywordeditsave"
        defaultMessage="Save"
        id="be.save"
      />
    </PrimaryButton>
  </div>
</span>
`;
